crates/agent/src/db/core.rs

```diff
@@ -2,5 +2,156 @@
 //!
 //! This module contains the core database types and functions.
 
-/// Database type
-pub type Database = ();
+use sqlx::migrate::MigrateError;
+use sqlx::postgres::{PgPool, PgPoolOptions};
+use std::time::Duration;
+use tracing::{error, info};
+
+/// Database type - PostgreSQL connection pool
+pub type Database = PgPool;
+
+/// Database error type
+pub type DbError = sqlx::Error;
+
+/// Custom error type for database operations
+#[derive(Debug, thiserror::Error)]
+pub enum DatabaseError {
+    /// Error from the underlying SQLx library
+    #[error("Database error: {0}")]
+    Sqlx(#[from] sqlx::Error),
+
+    /// Error during database migration
+    #[error("Migration error: {0}")]
+    Migration(#[from] MigrateError),
+
+    /// Other general errors
+    #[error("Other error: {0}")]
+    Other(String),
+}
+
+/// Extract database name from connection string
+fn extract_db_name(connection_params: &str) -> String {
+    connection_params
+        .split('/')
+        .next_back()
+        .and_then(|s| s.split('?').next())
+        .unwrap_or("wei_agent")
+        .to_string()
+}
+
+/// Check if the database exists and create it if it doesn't
+/// Returns a boolean indicating if the database was newly created
+/// Build a connection string to the postgres system database
+fn build_postgres_system_url(connection_params: &str) -> String {
```

Comment:
Would be nice to have it tested, but don't worry for now - we can do it in the future